File: custom-set/CustomSet.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class CustomSet
{
    private readonly ISet<int> set = new HashSet<int>();

    public CustomSet(params int[] values) =>
        values
            .OrderBy(value => value)
            .ToList()
            .ForEach(value => set.Add(value));

    public CustomSet Add(int value)
    {
        set.Add(value);
        return new CustomSet(set.ToArray());
    }

    public bool Empty()
        => !set.Any();

    public bool Contains(int value)
        => set.Contains(value);

    public bool Subset(CustomSet right)
        => set.IsSubsetOf(right.set);

    public bool Disjoint(CustomSet right)
        => !set.Any(value => right.set.Contains(value));

    public CustomSet Intersection(CustomSet right)
        => new CustomSet(set.Intersect(right.set).ToArray());

    public CustomSet Difference(CustomSet right)
        => new CustomSet(set.Except(right.set).ToArray());

    public CustomSet Union(CustomSet right)
        => new CustomSet(set.Union(right.set).ToArray());

    public override bool Equals(object obj)
        => obj is CustomSet customSet
            ? set.SequenceEqual(customSet.set)
            : false;

    public override int GetHashCode() => set.GetHashCode();
}

File: fsharp/custom-set/CustomSet.fs
module CustomSet

type Set<'T> = { Items: 'T list }

let empty =
    { Items = [] }

let singleton value =
    { Items = [value] }

let isEmpty set =
    set.Items.IsEmpty

let size set =
    set.Items.Length 

let fromList list =
    { 
        Items = list
            |> List.sort
            |> List.distinct 
    } 

let toList set =
    set.Items

let contains value set =
    List.contains value set.Items

let insert value set =
    value::set.Items
    |> fromList

let union left right =
    left.Items @ right.Items
    |> fromList

let intersection left right =
    left.Items
    |> List.filter (fun x -> List.contains x right.Items)
    |> fromList

let difference left right =
    left.Items
    |> List.filter (fun x -> List.contains x right.Items |> not)
    |> fromList

let isEqualTo left right =
    (size left = size right) && (isEmpty (difference left right))

let isSubsetOf left right =
    left.Items
    |> List.forall (fun x -> List.contains x right.Items)

let isDisjointFrom left right =
    left.Items
    |> List.exists (fun x -> List.contains x right.Items)
    |> not